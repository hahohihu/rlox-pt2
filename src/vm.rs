use std::{io::Write, ops::Range};

use ariadne::{Color, Label, Report, ReportKind, Source};

use crate::{
    chunk::{Chunk, OpCode},
    compiler::compile,
    object::Object,
    ui::{self, Span},
    value::Value,
};

struct VM<'src, Stderr, Stdout> {
    chunk: Chunk,
    ip: usize,
    stack: Vec<Value>,
    source: &'src str,
    stderr: Stderr,
    stdout: Stdout,
    // SAFETY INVARIANT: All objects in objects are valid, and there are no duplicate allocations
    objects: Vec<Object>,
}

impl<'src, Stderr, Stdout> Drop for VM<'src, Stderr, Stdout> {
    fn drop(&mut self) {
        for object in &self.objects {
            unsafe {
                // SAFETY: See safety invariant on objects
                object.free();
            }
        }
    }
}

#[derive(Debug)]
pub enum InterpretError {
    CompileError = 1,
    RuntimeError = 2,
}

type InterpretResult = Result<(), InterpretError>;

impl<'src, Stderr: Write, Stdout: Write> VM<'src, Stderr, Stdout> {
    fn new(chunk: Chunk, source: &'src str, stderr: Stderr, stdout: Stdout) -> Self {
        Self {
            chunk,
            source,
            stack: vec![],
            ip: 0,
            objects: vec![],
            stderr,
            stdout,
        }
    }

    fn next_byte(&mut self) -> u8 {
        self.ip += 1;
        self.chunk.instructions[self.ip - 1]
    }

    /// This gets the span over the range relative to the current IP
    /// Note that this is a bit wonky since instructions aren't in the same order as code
    /// (and are variable-length)
    fn get_span(&self, range: Range<isize>) -> Span {
        let start = self.ip as isize + range.start;
        let end = self.ip as isize + range.end;
        debug_assert!(start < end);
        if start < 0 {
            // The bytecode will always have at least a return
            return self.chunk.spans[0];
        }
        let range = Range {
            start: start.max(0) as usize,
            end: (end as usize).min(self.chunk.spans.len()),
        };
        Span::unite_many(&self.chunk.spans[range])
    }

    fn read_constant(&mut self) -> Value {
        let i = self.next_byte();
        self.chunk.get_constant(i)
    }

    fn binary_num_op(&mut self, name: &str, op: impl Fn(f64, f64) -> Value) -> InterpretResult {
        let b = self.stack.pop().unwrap();
        let a = self.stack.pop().unwrap();
        match (a, b) {
            (Value::Num(a), Value::Num(b)) => self.stack.push(op(a, b)),
            (a, b) => {
                let span = self.get_span(-2..1);
                self.runtime_error(
                    span,
                    format!(
                        "Operator '{name}' takes two numbers. Got a {} ({a}) and a {} ({b}).",
                        a.typename(),
                        b.typename()
                    ),
                );
                return Err(InterpretError::RuntimeError);
            }
        }
        Ok(())
    }

    fn runtime_error(&mut self, span: Span, message: String) {
        Report::build(ReportKind::Error, (), ui::OFFSET)
            .with_message(message)
            .with_label(Label::new(span).with_color(Color::Red))
            .finish()
            .write(Source::from(self.source), &mut self.stderr)
            .unwrap();
    }

    fn run(&mut self) -> InterpretResult {
        if self.chunk.instructions.is_empty() {
            return Ok(());
        }
        loop {
            #[cfg(feature = "verbose_vm")]
            {
                self.chunk
                    .disassemble_instruction(self.ip, self.source, std::io::stdout());
                println!("==== STACK ====");
                let stack_len = self.stack.len().saturating_sub(8);
                for value in &self.stack[stack_len..] {
                    println!("{value}");
                }
                println!("==============================");
            }
            let instruction: OpCode = self.next_byte().into();
            match instruction {
                OpCode::Return => {
                    return Ok(());
                }
                OpCode::Pop => {
                    self.stack.pop().expect("ICE: bad codegen");
                }
                OpCode::DefineGlobal => {
                    todo!()
                }
                OpCode::Constant => {
                    let constant = self.read_constant();
                    self.stack.push(constant);
                }
                OpCode::Nil => {
                    self.stack.push(Value::Nil);
                }
                OpCode::True => self.stack.push(Value::Bool(true)),
                OpCode::False => self.stack.push(Value::Bool(false)),
                OpCode::Negate => {
                    let val = self.stack.pop().unwrap();
                    match val {
                        Value::Num(n) => {
                            self.stack.push(Value::Num(-n));
                        }
                        val => {
                            let span = self.get_span(-3..0);
                            self.runtime_error(
                                span,
                                format!("Tried to negate a {} ({val})", val.typename()),
                            );
                            return Err(InterpretError::RuntimeError);
                        }
                    }
                }
                OpCode::Not => {
                    let value = Value::Bool(self.stack.pop().unwrap().falsey());
                    self.stack.push(value);
                }
                OpCode::Print => {
                    let value = self.stack.pop().unwrap();
                    writeln!(self.stdout, "{value}").unwrap();
                }
                OpCode::Add => {
                    let b = self.stack.pop().unwrap();
                    let a = self.stack.pop().unwrap();
                    match (a, b) {
                        (Value::Num(a), Value::Num(b)) => self.stack.push(Value::Num(a + b)),
                        (Value::Object(a), Value::Object(b)) if a.is_string() && b.is_string() => {
                            let concatenated = Object::concatenate_strings(a, b);
                            self.objects.push(concatenated);
                            self.stack.push(Value::Object(concatenated));
                        }
                        (a, b) => {
                            let span = self.get_span(-2..1);
                            self.runtime_error(
                                span,
                                format!(
                                    "Operator '+' takes two numbers. Got a {} ({a}) and a {} ({b}).",
                                    a.typename(),
                                    b.typename()
                                ),
                            );
                            return Err(InterpretError::RuntimeError);
                        }
                    }
                    Ok(())
                }?,
                OpCode::Sub => self.binary_num_op("-", |a, b| Value::Num(a - b))?,
                OpCode::Mul => self.binary_num_op("*", |a, b| Value::Num(a * b))?,
                OpCode::Div => self.binary_num_op("/", |a, b| Value::Num(a / b))?,
                OpCode::Less => self.binary_num_op("<", |a, b| Value::Bool(a < b))?,
                OpCode::Greater => self.binary_num_op(">", |a, b| Value::Bool(a > b))?,
                OpCode::Equal => {
                    let b = self.stack.pop().unwrap();
                    let a = self.stack.pop().unwrap();
                    self.stack.push(Value::Bool(a == b));
                }
                OpCode::Invalid => unreachable!("Reached invalid opcode at {}", self.ip),
            }
        }
    }
}

pub fn interpret(source: &str, mut stderr: impl Write, mut stdout: impl Write) -> InterpretResult {
    let chunk = match compile(source, &mut stderr) {
        Ok(chunk) => chunk,
        Err(e) => {
            e.print(&mut stderr, source);
            return Err(InterpretError::CompileError);
        }
    };
    let mut vm = VM::new(chunk, source, &mut stderr, &mut stdout);
    vm.run()
}

#[cfg(test)]
mod tests {
    use crate::snap;
    snap!(mismatched_add, "print true + 1;");
    snap!(mismatched_sub, "print true - 1;");
    snap!(negate_bool, "print -true;");
    snap!(one, "print 1;");
    snap!(point_one, "print 0.1;");
    snap!(print_false, "print false;");
    snap!(print_nil, "print nil;");
    snap!(add_mul, "print 1 + 2 * 3;");
    snap!(mul_div, "print 6 * 6 / 3;");
    snap!(complex_arithmetic, "print 20 * 5 / 0.5 - 100.0;");
    snap!(div_0, "print 1 / 0;");
    snap!(parens, "print 2 * (6 + 1) / (2) -- 100;");
    snap!(nested_parens, "print ((1) / (1 + (1 / 0.5)) * 3);");
    snap!(unary, "print -1 - -2 == --1 == true;");

    snap!(
        falsey,
        "
        print !nil;
        print !false;
        print !0;
        print !true;
        print !\"\";
    "
    );

    snap!(less_than, "print 1 < 1; print 0 < 1; print 2 < 1;");
    snap!(less_equal, "print 1 <= 1; print 0 <= 1; print 2 <= 1;");
    snap!(equality, "print 1 == 1; print 0 == 1; print 2 == 1;");
    snap!(greater_than, "print 1 > 1; print 0 > 1; print 2 > 1;");
    snap!(greater_equal, "print 1 >= 1; print 0 >= 1; print 2 >= 1;");

    snap!(string, r#"print "foo";"#);
    snap!(concatenation, r#"print "foo" + "bar";"#);
    snap!(string_comparison, r#"print "foo" == "foo";"#);
    snap!(
        compound_string_comparison,
        r#"print "foo" + "bar" == "f" + "oo" + "bar";"#
    );
    snap!(
        unicode,
        r#"print "💩" + "👪" + "༕" + "갍" + "⑯" + "ฒ" + "ڦ";"#
    );
}
