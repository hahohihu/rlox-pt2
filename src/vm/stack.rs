use std::{cell::UnsafeCell, mem::transmute};

use crate::repr::value::Value;

const MAX_SIZE: usize = 4096;
#[derive(Debug)]
pub struct FixedStack {
    // This abomination is necessary because of aliasing rules
    // In particular, we cannot take a &mut because we have pointers into the stack that mutate it
    // MaybeUninit is mostly an optimization, since we already have len
    stack: [UnsafeCell<Value>; MAX_SIZE],
    len: UnsafeCell<usize>,
}

impl FixedStack {
    pub fn new() -> Self {
        // this is just used to initialize the array, and is not actually mutated
        #[allow(clippy::declare_interior_mutable_const)]
        const UNINIT: UnsafeCell<Value> = UnsafeCell::new(Value::Num(f64::MAX));
        Self {
            stack: [UNINIT; MAX_SIZE],
            len: UnsafeCell::new(0),
        }
    }

    pub fn len(&self) -> usize {
        unsafe { *self.len.get() }
    }

    #[must_use]
    pub fn push(&self, value: Value) -> bool {
        unsafe {
            let len = self.len();
            if len >= MAX_SIZE {
                return false;
            }

            *self.stack[len].get() = value;
            *self.len.get() += 1;
            true
        }
    }

    #[must_use]
    pub fn peek(&self, i: usize) -> Option<Value> {
        unsafe {
            let len = self.len();
            let i = len - i - 1;
            if i >= len {
                return None;
            }
            Some(*self.get_ptr(i))
        }
    }

    pub fn get(&self, i: usize) -> Option<Value> {
        if i >= self.len() {
            None
        } else {
            unsafe { Some(*self.get_ptr(i)) }
        }
    }

    pub fn pop(&self) -> Option<Value> {
        let res = self.peek(0);
        if let Some(_) = res {
            unsafe {
                *self.len.get() -= 1;
            }
        }
        res
    }

    pub unsafe fn get_ptr(&self, index: usize) -> *mut Value {
        // MaybeUninit is repr(transparent)
        self.stack[index].get()
    }

    /// SAFETY: The burden of following aliasing rules is on the callee
    ///         With great power comes great responsibility
    pub unsafe fn slice(&self) -> &[Value] {
        // UnsafeCell is repr(transparent)
        transmute(&self.stack[..self.len()])
    }
}
