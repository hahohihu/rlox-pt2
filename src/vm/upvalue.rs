use std::fmt::Display;

use crate::value::{valid::ValidPtr, Value};

#[derive(Copy, Clone, Debug)]
pub struct Upvalue {
    pub value: ValidPtr<Value>,
    /// This should never be directly accessed after initialization, it's just storage
    closed: Value,
    pub marked: bool,
    pub next_open: Option<ValidPtr<Upvalue>>,
}

impl Upvalue {
    pub fn new(value: ValidPtr<Value>, next_open: Option<ValidPtr<Upvalue>>) -> Self {
        Self {
            value,
            closed: Value::Num(f64::MAX),
            marked: false,
            next_open,
        }
    }

    pub fn close(my: ValidPtr<Self>) {
        unsafe {
            /*
            This is deeply unsound under stack borrows, since it pops off the &mut VM that calls it
            But that VM only ever references the stack immutably, so that seems rather silly
            Tree borrows are ok with this, and hopefully whatever model Rust winds up with does too
            Regardless, this is a bit of a wont-fix. Fixing it would require making the VM extremely un-ergonomic
            And this is a learning project after all.
             */
            (*my.as_ptr()).closed = *my.value;
            (*my.as_ptr()).value = ValidPtr::from_ptr(&mut (*my.as_ptr()).closed);
        }
    }

    pub fn mark(&self) {
        self.closed.mark();
    }
}

impl PartialEq for Upvalue {
    fn eq(&self, other: &Self) -> bool {
        self.value.as_ptr() == other.value.as_ptr()
    }
}

impl Eq for Upvalue {}

impl Display for Upvalue {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "<upvalue = {} @ {:?}>", *self.value, self.value.as_ptr())
    }
}
