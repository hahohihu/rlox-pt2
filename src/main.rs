#![allow(clippy::redundant_pattern_matching)]
use std::{
    env::args,
    fs::File,
    io::{stderr, Read},
    process::ExitCode,
};

use vm::interpret;

mod chunk;
mod lex;
mod object;
mod parse;
mod ui;
mod util;
mod valid;
mod value;
mod vm;

fn read_file(filename: &str) -> std::io::Result<String> {
    let mut file = File::open(filename)?;
    let mut source = String::new();
    file.read_to_string(&mut source)?;
    Ok(source)
}

fn main() -> ExitCode {
    tracing_subscriber::fmt::init();
    let mut args = args();
    args.next();
    let Some(filename) = args.next() else {
        eprintln!("Usage: rlox <filename>");
        return ExitCode::FAILURE;
    };
    let source = match read_file(&filename) {
        Ok(file) => file,
        Err(e) => {
            eprintln!("Failed to read {filename} with error: {:?}", e);
            return ExitCode::FAILURE;
        }
    };
    match interpret(&source, stderr()) {
        Ok(_) => ExitCode::SUCCESS,
        Err(_) => ExitCode::FAILURE,
    }
}
