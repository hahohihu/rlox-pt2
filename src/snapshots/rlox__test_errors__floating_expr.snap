---
source: src/main.rs
expression: "crate::util::mock_interpret(\"1;\")"
---
stdout:


stderr:


